src/Artifacts.jl

```diff
@@ -0,0 +1,594 @@
+most appropriate mapping.  If none is found, return `nothing`.
+"""
+function artifact_meta(name::String, artifacts_toml::String;
+                       platform::Platform = platform_key_abi())
+    if !isfile(artifacts_toml)
+        return nothing
+    end
+
+    # Parse the toml for the 
+    return artifact_meta(name, parse_toml(artifacts_toml), artifacts_toml; platform=platform)
+end
+
+function artifact_meta(name::String, artifact_dict::Dict, artifacts_toml::String;
+                       platform::Platform = platform_key_abi())
+    if !haskey(artifact_dict, name)
+        return nothing
+    end
+    meta = artifact_dict[name]
+
+    # If it's an array, find the entry that best matches our current platform
+    if isa(meta, Array)
+        dl_dict = Dict(unpack_platform(x, name, artifacts_toml) => x for x in meta)
+        meta = select_platform(dl_dict, platform)
+    # If it's NOT a dict, complain
+    elseif !isa(meta, Dict)
+        @warn("Invalid Artifacts.toml at $(artifacts_toml): artifact '$name' malformed, must be array or dict!")
+        return nothing
+    end
+
+    # This is such a no-no, we are going to call it out right here, right now.
+    if meta != nothing && !haskey(meta, "git-tree-sha1")
+        @warn("Invalid Artifacts.toml at $(artifacts_toml): artifact '$name' contains no `git-tree-sha1`!")
+        return nothing
+    end
+
+    # Return the full meta-dict.
+    return meta
+end
+
+"""
+    artifact_hash(name::String, artifacts_toml::String; platform::Platform = platform_key_abi())
+
+Thin wrapper around `artifact_meta()` to return the hash of the specified, platform-
+collapsed artifact.  Returns `nothing` if no mapping can be found.
+"""
+function artifact_hash(name::String, artifacts_toml::String;
+                       platform::Platform = platform_key_abi())
+    meta = artifact_meta(name, artifacts_toml; platform=platform)
+    if meta === nothing
+        return nothing
+    end
+
+    return SHA1(meta["git-tree-sha1"])
+end
+
+"""
+    bind_artifact(name::String, hash::SHA1, artifacts_toml::String;
+                  platform::Union{Platform,Nothing} = nothing,
+                  download_info::Union{Vector{Tuple},Nothing} = nothing,
+                  force::Bool = false)
+
+Writes a mapping of `name` -> `hash` within the given `Artifacts.toml` file.  If
+`platform` is not `nothing`, this artifact is marked as platform-specific, and will be
+a multi-mapping.  It is valid to bind multiple artifacts with the same name, but
+different `platform`s and `hash`'es within the same `artifacts_toml`.  If `force` is set
+to `true`, this will overwrite a pre-existant mapping, otherwise an error is raised.
+
+`download_info` is an optional tuple that contains a vector of URLs and a hash.  These
+URLs will be listed as possible locations where this artifact can be obtained.  The
+"""
+function bind_artifact(name::String, hash::SHA1, artifacts_toml::String;
```

Comment:
I wonder if this should be 
`write_artifact_binding` 
so something, to make it obvious that it writes to disk

Reply:
I hope that needing to pass in a filename helps get that across.  I am willing to go as far as re-arranging this to be:

```
bind_artifact!(artifacts_toml, name, hash)
```

If you think that is clearer.

Reply:
Sold